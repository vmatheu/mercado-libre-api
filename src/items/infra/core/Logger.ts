import { Injectable, Logger, LoggerService, Scope } from '@nestjs/common';

@Injectable({ scope: Scope.TRANSIENT })
export class LoggerCustom implements LoggerService {
  private prefix: string;
  private logger: Logger;

  init(prefix: string) {
    this.logger = new Logger(prefix);
  }

  log(message: any, ...optionalParams: any[]) {
    this.logger.log(message);
  }
  error(message: any, ...optionalParams: any[]) {
    this.logger.error(message);
  }
  warn(message: any, ...optionalParams: any[]) {
    this.logger.warn(message);
  }
  debug?(message: any, ...optionalParams: any[]) {
    this.logger.debug(message);
  }
  verbose?(message: any, ...optionalParams: any[]) {
    this.logger.verbose(message);
  }
}
