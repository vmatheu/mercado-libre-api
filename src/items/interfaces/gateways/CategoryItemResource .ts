export interface CategoryItemResource {
  getCategories: (categoryId: string) => Promise<string[]>;
}
