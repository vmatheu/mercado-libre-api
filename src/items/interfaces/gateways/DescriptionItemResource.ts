export interface DescriptionItemResource {
  getDescription: (id: string) => Promise<string>;
}
