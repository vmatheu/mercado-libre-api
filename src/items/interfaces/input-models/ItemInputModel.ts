export interface ItemInputModel {
  id: string;
  title: string;
  price: number;
  currencyId: string;
  condition: string;
  picture: string;
  freeShipping: boolean;
  category: string;
}
