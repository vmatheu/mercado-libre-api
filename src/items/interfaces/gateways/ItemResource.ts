import { ItemWithSoldInputModel } from '../input-models/ItemWithSoldInputModel';

export interface ItemResource {
  getById: (id: string) => Promise<ItemWithSoldInputModel>;
}
