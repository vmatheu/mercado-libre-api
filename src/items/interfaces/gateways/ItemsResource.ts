import { ItemInputModel } from '../input-models/ItemInputModel';

export interface ItemsResource {
  findByQuerySearch: (query: string) => Promise<ItemInputModel[]>;
}
