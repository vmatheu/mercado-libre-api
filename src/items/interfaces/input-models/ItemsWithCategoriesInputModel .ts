import { ItemInputModel } from "./ItemInputModel";

export interface ItemWithCategoriesInputModel {
  items: ItemInputModel[]
  categories: string []
}
