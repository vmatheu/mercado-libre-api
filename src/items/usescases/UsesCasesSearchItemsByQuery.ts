import { Injectable } from '@nestjs/common';
import { buildItem } from '../domain/Item';
import { LoggerCustom } from '../infra/core/Logger';
import { ItemsResourcesEndpoint } from '../infra/endpoints/ItemsResourcesEndpoint';
import { ItemInputModel } from '../interfaces/input-models/ItemInputModel';
import { QueryInputModel } from '../interfaces/input-models/QueryInputModel';
import {
  Item,
  SearchOutputModel,
} from '../interfaces/output-models/SearchOutputModel';

interface IHash {
  [details: string]: number;
}

@Injectable()
export class UsesCasesSearchItemsByQuery {
  constructor(
    private readonly itemsResource: ItemsResourcesEndpoint,
    private readonly logger: LoggerCustom,
  ) {
    logger.init(UsesCasesSearchItemsByQuery.name);
  }

  async findByQuerySearch(
    request: QueryInputModel,
  ): Promise<SearchOutputModel> {
    this.logger.debug(`call findByQuerySearch=${request.q}`);

    const result = await this.itemsResource.findByQuerySearch(request.q);
    const categoryMap: IHash = {};
    const items = result.slice(0,4).map(transformInputSearchToItems(categoryMap));
    const categories = Object.keys(categoryMap).map((key) => key);

    this.logger.debug(`result findByQuerySearch count=${items.length}`);

    return {
      author: {
        name: 'Victor',
        lastName: 'Matheu',
      },
      categories,
      items,
    };
  }
}

function transformInputSearchToItems(
  categoryMap: IHash,
): (value: ItemInputModel, index: number, array: ItemInputModel[]) => Item {
  return (item) => {
    categoryMap[item.category] = 1;
    return buildItem(item);
  };
}
