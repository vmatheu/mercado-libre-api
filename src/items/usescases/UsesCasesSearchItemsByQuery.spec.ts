import { ItemsResource } from '../interfaces/gateways/ItemsResource';
import { UsesCasesSearchItemsByQuery } from './UsesCasesSearchItemsByQuery';
import { mock, mockReset } from 'jest-mock-extended';
import { LoggerCustom } from '../infra/core/Logger';
import { ItemInputModel } from '../interfaces/input-models/ItemInputModel';

describe('UsesCasesSearchItemById', () => {
  const itemsResourceMock = mock<ItemsResource>();
  const loggerMock = mock<LoggerCustom>();

  const itemMock: ItemInputModel = {
    id: '123231',
    title: 'Test',
    price: 1009,
    currencyId: 'USD',
    condition: 'new',
    picture: 'no ttiene',
    freeShipping: false,
  };

  beforeEach(async () => {
    mockReset(itemsResourceMock);
  });

  it('should return result item object when call getItemByI', async () => {
    itemsResourceMock.findByQuerySearch.mockResolvedValue({
      items: [itemMock],
      categories: ['categoria']
    });

    const result = await new UsesCasesSearchItemsByQuery(
      itemsResourceMock,
      loggerMock,
    ).findByQuerySearch({
      q: 'mi busqueda',
    });

    expect(result).toMatchSnapshot();
  });
});
