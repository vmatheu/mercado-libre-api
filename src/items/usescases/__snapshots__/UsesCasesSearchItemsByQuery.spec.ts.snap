// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsesCasesSearchItemById should return result item object when call getItemByI 1`] = `
Object {
  "author": Object {
    "lastName": "Matheu",
    "name": "Victor",
  },
  "categories": Array [
    "Mejor",
  ],
  "items": Array [
    Object {
      "condition": "new",
      "freeShipping": false,
      "id": "123231",
      "picture": "no ttiene",
      "price": Object {
        "amount": 1009,
        "currency": "USD",
        "decimals": 0,
      },
      "title": "Test",
    },
  ],
}
`;
